File: src/cleanerversion/SparseHashedMatrix.java
package cleanerversion;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class SparseHashedMatrix implements Matrix {
	private List<Map<Integer, Float>> values = null;
	private int N, M;

	public SparseHashedMatrix(int n, int m) {
		N = n;
		M = m;

		values = new ArrayList<Map<Integer, Float>>(N);
		for (int i=0; i<N; i++) {
			values.add(i, new HashMap<Integer, Float>());
		}
	}

	@Override
	public Matrix clone() {
		Matrix clone = new SparseHashedMatrix(N, M);

		for (int row=0; row<N; row++) {
			for (int column : getKeysOfNonZeroElementsForRow(row)) {
				clone.setValue(column, row, getValue(row, column));
			}
		}
		
		return clone;
	}

	private Collection<Integer> getKeysOfNonZeroElementsForRow(int row) {
		return values.get(row).keySet();
	}

	@Override
	public int getM() {
		return M;
	}

	@Override
	public int getN() {
		return N;
	}

	@Override
	public float getValue(int column, int row) {
		return values.get(row).get(column);
	}

	@Override
	public void setValue(int column, int row, float value) {
		values.get(row).put(column, value);
	}

	@Override
	public Vector multiply(Vector v) {
		if (v.getSize() != this.getM()) {
			throw new IllegalArgumentException("The vector must be of the " +
					"same length as the horizontal length of this matrix");
		}

		Vector result = new ArrayVector(v.getSize());

		// TODO: this should be threaded
		for (int row=0; row<N; row++) {
			float temp = 0.0f;
			for (int column : getKeysOfNonZeroElementsForRow(row)) {
				temp += this.getValue(column, row) * v.getValue(column);
			}
			result.setValue(row, temp);
		}

		return result;
	}

	@Override
	public void printToMatrixMarketFile(String fileName) throws IOException {
		printToMatrixMarketFile(fileName, false);
	}

	@Override
	public void printToMatrixMarketFile(String fileName, boolean decrementCount) throws IOException {
		int numEntries = 0;
		for (int i=0; i<this.getM(); i++) {
			numEntries += this.getKeysOfNonZeroElementsForRow(i).size();
		}
		int mCount = this.getM();
		int nCount = this.getN();
		if (decrementCount) {
			mCount -= 1;
			nCount -= 1;
		}

		BufferedWriter out = new BufferedWriter(new FileWriter(fileName + ".mtx"));
		out.write("%%MatrixMarket matrix coordinate real general");
		out.newLine();
		out.write("%");
		out.newLine();
		out.write(mCount + " " + nCount + " " + numEntries);
		out.newLine();
		for (int row=0; row<this.getN(); row++) {
			for (int column : this.getKeysOfNonZeroElementsForRow(row)) {
				if (this.getValue(column, row) != 0.0f) {
					out.write(row + " " + column + " " + this.getValue(column, row));
					out.newLine();
				}
			}
		}
		out.close();
	}
}

File: cpp/sparse_hashed_matrix.cpp
#include "matrix.h"
#include "vector.h"
#include "matrix_market_handler.h"
#include <iostream>
#include <map>
#include <iterator>
#include <omp.h>

using namespace std;

template <typename T>
class SparseHashedMatrix : public Matrix<T> {
private:
	map<unsigned int, T> *values;
	unsigned int M;
	unsigned int N;

public:
	SparseHashedMatrix<T>() {
		M = 0;
		N = 0;
	}

	SparseHashedMatrix<T>(unsigned int m, unsigned int n) {
		M = m;
		N = n;
		values = new map<unsigned int, T>[N];
	}

	SparseHashedMatrix<T>(const SparseHashedMatrix<T> &m2) {
		M = m2.getM();
		N = m2.getN();

		typename map<unsigned int, T>::iterator itr;
		for (unsigned int row=0; row<getN(); row++) {
			for (itr = m2.values[row].begin(); itr != m2.values[row].end(); ++itr) {
				setValue(itr->first, row, itr->second);
			}
		}
	}

	SparseHashedMatrix<T>(string fileName) {
		M = 0;
		N = 0;
		readFromMatrixMarketFile(fileName);
	}

	~SparseHashedMatrix<T>() {
		delete[] values;
	}

	unsigned int getM() const {
		return M;
	}

	unsigned int getN() const {
		return N;
	}

	T getValue(unsigned int column, unsigned int row) const {
		if (values[row].find(column) == values[row].end()) {
			return 0;
		} else {
			return values[row].find(column)->second;
		}
	}

	void setValue(unsigned int column, unsigned int row, T value) {
		values[row][column] = value;
	}

	void multiply(Vector<T> &v, Vector<T> &result) const {
		int nprocs = sysconf(_SC_NPROCESSORS_ONLN);
		//cout << "using " << nprocs << " threads" << endl;
		omp_set_num_threads(nprocs);

		typename map<unsigned int, T>::iterator itr;
		#pragma omp parallel for private(itr) // probably should use blocking
		for (unsigned int row=0; row<getN(); row++) {
			T temp = 0;
			for (itr = values[row].begin(); itr != values[row].end(); ++itr) {
				temp += v.getValue(itr->first) * itr->second;
			}
			result.setValue(row, temp);
		}
	}

	void printToMatrixMarketFile(string fileName) {
		unsigned int numEntries = 0;

		typename map<unsigned int, T>::iterator itr;

		for (unsigned int row=0; row<getN(); row++) {
			for (itr = values[row].begin(); itr != values[row].end(); ++itr) {
				numEntries++;
			}
		}

		ofstream mmFile (fileName.c_str());

		if (mmFile.is_open()) {
			mmFile << "%%MatrixMarket matrix coordinate real general" << endl;
			mmFile << "%" << endl;
			mmFile << getM()-1 << " " << getN()-1 << " " << numEntries << endl;

			for (unsigned int row=0; row<getN(); row++) {
				for (itr = values[row].begin(); itr != values[row].end(); ++itr) {
					mmFile << row << " " << itr->first << " " << itr->second << endl;
				}
			}

			mmFile.close();
		} else {
			cout << "Problem writing matrix market file" << endl;
		}
	}

	// seriously don't use this except when testing
	// with large matrices, things will go kaplooy
	void print() {
		for (unsigned int row=0; row<getN(); row++) {
			for (unsigned int column=0; column<getM(); column++) {
				cout << " " << getValue(column, row);
			}
			cout << endl;
		}
	}

	void readFromMatrixMarketFile(string fileName) {
		unsigned int row, col;
		T val; // really this has to be double or float

		string line;
		ifstream mmFile (fileName.c_str());

		if (mmFile.is_open()) {
			while (mmFile.good()) {
				getline(mmFile, line);

				if (line[0] == '%') {
					continue;
				}

				std::stringstream lineStream(line);

				lineStream >> row >> col  >> val;


				if (M == 0 && N == 0) { // not allocated yet
					N = row + 1;
					M = col + 1;
					values = new map<unsigned int, T>[N];
					cout << "Matrix size: " << M << ", " << N << endl;

					continue;
				}

				// Not too sure if I should swap row with col
				setValue(col, row, val);

				// TODO : remove. for checking only
				//cout << "Setting val: col " << col << ", row" << row << " val " << val << endl;
				//cout << getValue(col, row) << endl;
			}

			mmFile.close();
		}

	}


};
